[WTF] Narrow scope of ignored -Wunsafe-buffer-usage warnings from ASCIILiteral.h to SuperFastHash.h

ddkilzer opened the pull request:
#### 00e4ae32b87a36717be7ed03a9f62cf6bdc456ce
<pre>
[WTF] Narrow scope of ignored -Wunsafe-buffer-usage warnings from ASCIILiteral.h to SuperFastHash.h
&lt;<a href="https://bugs.webkit.org/show_bug.cgi?id=281506">https://bugs.webkit.org/show_bug.cgi?id=281506</a>&gt;
&lt;<a href="https://rdar.apple.com/137973769">rdar://137973769</a>&gt;

Reviewed by Geoffrey Garen.

Additional changes in JavaScriptCore and WebKitTestRunner were due to
temporarily introducing an overloaded ASCIILiteral::fromLiteralUnsafe()
with a second size_t argument, which was later removed.  This caused a
disambiguation build failure on gcc, but these changes will be needed
later anyway, so they were kept.

* Source/WTF/wtf/text/ASCIILiteral.h:
- Move WTF_ALLOW_UNSAFE_BUFFER_USAGE_{BEGIN,END} macros to
  SuperFastHash.h.
(WTF::ASCIILiteral::fromLiteralUnsafe):
- Move call to std::char_traits&lt;char&gt;::length() into this method from
  the ASCIILiteral constructor.
- Construct a std::span that includes NUL terminator byte using
  unsafeForgeSpan() now that the ASCIILiteral constructor takes a
  std::span.
(WTF::ASCIILiteral::span8):
- Use std::span::first() to fix unsafe buffer usage warning.
(WTF::ASCIILiteral::ASCIILiteral):
- Change to take a std::span argument that includes the NUL terminator
  byte.
- Fix indentation of code in ASSERT_ENABLED.
(WTF::StringLiterals::operator&quot;&quot; _span):
- Create std::span first, then use that to check characters in Debug
  builds. This fixes an unsafe buffer usage warning.
- Use unsafeForgeSpan() to silence false positive warning.
* Source/WTF/wtf/StdLibExtras.h:
(WTF::spanReinterpretCast):
- Make constexpr by updating release assertion.
(WTF::unsafeForgeSpan):
- Make constexpr.
* Source/WTF/wtf/text/SuperFastHash.h:
- Move WTF_ALLOW_UNSAFE_BUFFER_USAGE_{BEGIN,END} macros to this header.

* Source/JavaScriptCore/runtime/IntlCollator.cpp:
(JSC::IntlCollator::sortLocaleData):
* Source/JavaScriptCore/runtime/IntlDateTimeFormat.cpp:
(JSC::IntlDateTimeFormat::localeData):
* Source/JavaScriptCore/runtime/IntlLocale.cpp:
(JSC::IntlLocale::calendars):
(JSC::IntlLocale::collations):
* Source/JavaScriptCore/runtime/IntlObject.cpp:
(JSC::intlAvailableCalendars):
(JSC::availableCollations):
(JSC::availableCurrencies):
(JSC::intlAvailableTimeZones):
* Source/JavaScriptCore/runtime/IntlPluralRules.cpp:
(JSC::IntlPluralRules::resolvedOptions const):
* Source/JavaScriptCore/runtime/JSONObject.cpp:
(JSC::gap):
- Construct std::span safely from C-array, then use first().
* Source/JavaScriptCore/tools/FunctionAllowlist.cpp:
(JSC::FunctionAllowlist::FunctionAllowlist):
* Source/JavaScriptCore/tools/FunctionOverrides.cpp:
(JSC::parseClause):
- Make use of unsafeForgeSpan(), except where noted above.  Many of
  these are after a call to uenum_next().

* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::createTestURL):
- Convert std::string directly to std::span.

Canonical link: <a href="https://commits.webkit.org/285440@main">https://commits.webkit.org/285440@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a7bbc5517e948bc97dca09826dccd6692613cce0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72625 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52050 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25423 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76814 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/23841 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74740 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59855 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23659 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/57144 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/23841 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75692 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47097 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62534 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/37571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43747 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20001 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22188 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/65758 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65605 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20360 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78484 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/71881 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16871 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19490 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16919 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62543 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/13164 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6809 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/93660 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11157 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47848 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20614 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48915 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/50210 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48660 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren reviewed (approved):
r=me

Now that we're narrowing enforcement scope, we'll soon need a separate macro to distinguish "this is skipped because it's safe" vs "this is skipped because we haven't gotten to it yet".

geoffreygaren commented on Source/WTF/wtf/text/ASCIILiteral.h:
```diff
@@ -145,17 +141,16 @@ constexpr ASCIILiteral operator"" _s(const char* characters, size_t)
 
 constexpr std::span<const LChar> operator"" _span(const char* characters, size_t n)
 {
+    auto span = std::span { bitwise_cast<const LChar*>(characters), n };
```
I'm surprised the compiler didn't complain here about using the two argument span constructor. But maybe it's smart enough to know the provenance of the 'n' passed to operator""?

If the compiler does ultimately complain about this, we'll need to add the unsafeForgeSpan function I described in Slack.

ddkilzer replied:
I just realized that I'm using an older clang compiler from macOS 14.x, not one from macOS 15, so the newer clang may find more issues.  (This is going to be an issue if EWS bots aren't updated to macOS 15 soon.)

I may need to abandon this PR since I was trying to do a small fix, but it's turning into a bigger fix.  :(

ddkilzer replied:
Nevermind, I tested on macOS 15 and have a fix.

geoffreygaren reviewed (approved):
r=me

Looks like the right approach; some bots are still red.

ddkilzer commented on Source/WTF/wtf/text/ASCIILiteral.h:
```diff
@@ -143,19 +140,20 @@ constexpr ASCIILiteral operator"" _s(const char* characters, size_t)
 {
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
+    auto span = spanReinterpretCast<const LChar>(std::span { characters, std::char_traits<char>::length(characters) });
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
```
This warns on macOS 15, but shouldn't because it's safe.

@geoffreygaren do we want to add `unsafeForgeSpan()` here, or make a different macro to denote safe patterns?

geoffreygaren replied:
unsafeForgeSpan() is the only option here because the pointer and length come to us as separate values, through an API that we can't change.

geoffreygaren commented on Source/WTF/wtf/text/ASCIILiteral.h:
```diff
@@ -143,19 +140,18 @@ constexpr ASCIILiteral operator"" _s(const char* characters, size_t)
-constexpr std::span<const LChar> operator"" _span(const char* characters, size_t n)
+constexpr std::span<const LChar> operator"" _span(const char* characters, size_t)
 {
+    auto span = unsafeForgeSpan(bitwise_cast<const LChar*>(characters), std::char_traits<char>::length(characters));
```
I think this inserts a new call to strlen() when the length was otherwise known at compile time, right? I'd prefer to use 'n' here instead, for performance, and also it's kinda safer? Because we know it's the 'n' we got from the compiler's constant string.

geoffreygaren replied:
(I've suggested, separately to clang developers, that the warning should improve not even to warn about 'n' in this case, for the same reason.)

ddkilzer replied:
Oof.  If `std::char_traits<char>::length()` calls `strlen()` at runtime, we're paying that cost for all of our constant strings using the `_s` operator!

I can fix that, though.

geoffreygaren replied:
Just to be sure of my reading of the docs, I double-checked:

```
~/scratch> cat ./scratch.cc                                                               
#include <memory>
#include <array>
#include <span>
#include <string>

using namespace std;

size_t f(const char* characters)
{
	return std::char_traits<char>::length(characters);
}

int main()
{
}
~/scratch> clang++ -std=c++23 -O3 -o scratch scratch.cc  && otool -tvV ./scratch | c++filt
./scratch:
(__TEXT,__text) section
f(char const*):
0000000100003f90	b	0x100003f9c ; symbol stub for: _strlen
_main:
0000000100003f94	mov	w0, #0x0
0000000100003f98	ret

```

geoffreygaren replied:
In the constant case, if you get sufficient inlining, the call to strlen disappears. So not a disaster. But not optimal.

ddkilzer replied:
Latest PR fixed both `_span` as well as `_s` operators.

ddkilzer replied:
Of course, that fix triggered buggy behavior in gcc that can't decide between `String::String(std::span<const char>)` constructor and `String::String(ASCIILiteral)` with `ASCIILiteral::ASCIILiteral(std::span<const char>)` constructor, even though the `ASCIILiteral` constructor is `private` and not even available to use.

geoffreygaren reviewed (approved):
Looks good but I think we want to remove the new call to strlen().

ddkilzer commented on Source/WTF/wtf/text/ASCIILiteral.h:
```diff
@@ -136,26 +138,25 @@ inline ASCIILiteral ASCIILiteral::deletedValue()
+    auto result = ASCIILiteral::fromLiteralUnsafe(characters, n);
     ASSERT_UNDER_CONSTEXPR_CONTEXT(result.characters() == characters);
     return result;
 }
```
LOL WUT.

```
$ ./Tools/Scripts/run-api-tests --release --no-build TestWTF.WTF.StringViewEqualBasic 
Running tests       
                  
./Tools/TestWebKitAPI/Tests/WTF/StringView.cpp:327
Value of: a == "Hell\0"_s
  Actual: true
Expected: false



./Tools/TestWebKitAPI/Tests/WTF/StringView.cpp:331
Value of: test3 == "Hello\0"_s
  Actual: false
Expected: true


worker/0 TestWTF.WTF.StringViewEqualBasic Failed
Ran 1 tests of 1 with 0 successful
------------------------------
Test suite failed

Failed

    TestWTF.WTF.StringViewEqualBasic
        
        ./Tools/TestWebKitAPI/Tests/WTF/StringView.cpp:327
        Value of: a == "Hell\0"_s
          Actual: true
        Expected: false
        
        
        
        ./Tools/TestWebKitAPI/Tests/WTF/StringView.cpp:331
        Value of: test3 == "Hello\0"_s
          Actual: false
        Expected: true
```

```
$ ./Tools/Scripts/run-api-tests --release --no-build TestWebKitAPI.ParsedContentType.Serialize
Running tests       

./Tools/TestWebKitAPI/Tests/WebCore/ParsedContentType.cpp:196
Expected equality of these values:
  serializeIfValid("text/plain\0"_s)
    Which is: "NOTVALID"
  "text/plain"


worker/0 TestWebKitAPI.ParsedContentType.Serialize Failed
Ran 1 tests of 1 with 0 successful
------------------------------
Test suite failed

Failed

    TestWebKitAPI.ParsedContentType.Serialize
        
        ./Tools/TestWebKitAPI/Tests/WebCore/ParsedContentType.cpp:196
        Expected equality of these values:
          serializeIfValid("text/plain\0"_s)
            Which is: "NOTVALID"
          "text/plain"
```

Note that `_span` doesn't have this behavior currently @geoffreygaren.

We're paying the runtime cost of `strlen()` for every `_s` operator to catch programmer errors.

Is there a fancy, template-y way to catch the `\0` in the string at build time?

ddkilzer replied:
> Is there a fancy, template-y way to catch the \0 in the string at build time?

I figured out how to make one of these.  See [User-defined literals](https://en.cppreference.com/w/cpp/language/user_literal).  I'll include it in another patch.

ddkilzer replied:
Turns out user-defined literals treat the global C-string as an array of bytes (passed around via function argument or template parameter) , but they never allocate the C-string as a global, so they can't be used to check that the C-string itself is well-formed (for example, no NULL bytes and all `isASCII()` characters).

It's too bad there's no way in C++ to get both the C-string global and the user-defined literal, because the literal could be used for validation (with a `static_assert()` expression, and the global could be used for the actual bytes.

Commit:
[WTF] Narrow scope of ignored -Wunsafe-buffer-usage warnings from ASCIILiteral.h to SuperFastHash.h
<https://bugs.webkit.org/show_bug.cgi?id=281506>
<rdar://137973769>

Reviewed by Geoffrey Garen.

Additional changes in JavaScriptCore and WebKitTestRunner were due to
temporarily introducing an overloaded ASCIILiteral::fromLiteralUnsafe()
with a second size_t argument, which was later removed.  This caused a
disambiguation build failure on gcc, but these changes will be needed
later anyway, so they were kept.

* Source/WTF/wtf/text/ASCIILiteral.h:
- Move WTF_ALLOW_UNSAFE_BUFFER_USAGE_{BEGIN,END} macros to
  SuperFastHash.h.
(WTF::ASCIILiteral::fromLiteralUnsafe):
- Move call to std::char_traits<char>::length() into this method from
  the ASCIILiteral constructor.
- Construct a std::span that includes NUL terminator byte using
  unsafeForgeSpan() now that the ASCIILiteral constructor takes a
  std::span.
(WTF::ASCIILiteral::span8):
- Use std::span::first() to fix unsafe buffer usage warning.
(WTF::ASCIILiteral::ASCIILiteral):
- Change to take a std::span argument that includes the NUL terminator
  byte.
- Fix indentation of code in ASSERT_ENABLED.
(WTF::StringLiterals::operator"" _span):
- Create std::span first, then use that to check characters in Debug
  builds. This fixes an unsafe buffer usage warning.
- Use unsafeForgeSpan() to silence false positive warning.
* Source/WTF/wtf/StdLibExtras.h:
(WTF::spanReinterpretCast):
- Make constexpr by updating release assertion.
(WTF::unsafeForgeSpan):
- Make constexpr.
* Source/WTF/wtf/text/SuperFastHash.h:
- Move WTF_ALLOW_UNSAFE_BUFFER_USAGE_{BEGIN,END} macros to this header.

* Source/JavaScriptCore/runtime/IntlCollator.cpp:
(JSC::IntlCollator::sortLocaleData):
* Source/JavaScriptCore/runtime/IntlDateTimeFormat.cpp:
(JSC::IntlDateTimeFormat::localeData):
* Source/JavaScriptCore/runtime/IntlLocale.cpp:
(JSC::IntlLocale::calendars):
(JSC::IntlLocale::collations):
* Source/JavaScriptCore/runtime/IntlObject.cpp:
(JSC::intlAvailableCalendars):
(JSC::availableCollations):
(JSC::availableCurrencies):
(JSC::intlAvailableTimeZones):
* Source/JavaScriptCore/runtime/IntlPluralRules.cpp:
(JSC::IntlPluralRules::resolvedOptions const):
* Source/JavaScriptCore/runtime/JSONObject.cpp:
(JSC::gap):
- Construct std::span safely from C-array, then use first().
* Source/JavaScriptCore/tools/FunctionAllowlist.cpp:
(JSC::FunctionAllowlist::FunctionAllowlist):
* Source/JavaScriptCore/tools/FunctionOverrides.cpp:
(JSC::parseClause):
- Make use of unsafeForgeSpan(), except where noted above.  Many of
  these are after a call to uenum_next().

* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::createTestURL):
- Convert std::string directly to std::span.

Canonical link: https://commits.webkit.org/285440@main
